printers/vcd_witness_printer.cpp

```diff
@@ -0,0 +1,542 @@
+
+  std::map<std::string, VCDArray> & signal_set = root->arrays;
+
+  if (signal_set.find(short_name) != signal_set.end()) {
+    // TODO: only check in Debug
+    throw CosaException(full_name + " has been registered already");
+  }
+
+  signal_set.insert(std::make_pair(short_name,
+    VCDArray(short_name, full_name,  ast, data_width)));
+  auto & indices2hash = signal_set.at(short_name).indices2hash;
+  for (const auto & index : indices)
+    indices2hash.insert(std::make_pair(index, new_hash_id()));
+  if (has_default)
+    indices2hash.insert(std::make_pair("default", new_hash_id()));
+  allsig_array_.push_back( &(signal_set.at(short_name)) );
+  // to do: add indices and their hashes
+} // end of check_insert_scope_array
+
+
+
+void VCDWitnessPrinter::dump_current_scope(std::ostream & fout, const VCDScope *scope) const {
+  for (auto && r : scope->regs) {
+    fout << "$var reg " << r.second.data_width << " " << r.second.hash << " "
+         << r.second.vcd_name << " $end" << std::endl;
+  }
+  for (auto && w : scope->wires) {
+    fout << "$var wire " << w.second.data_width << " " << w.second.hash << " "
+         << w.second.vcd_name << " $end" << std::endl;
+  }
+  for (auto && a : scope->arrays) {
+    auto data_width = a.second.data_width;
+    for (auto && idx_hash_pair : a.second.indices2hash)
+      fout << "$var reg " << data_width << " " << idx_hash_pair.second << " "
+           << a.second.vcd_name+"[" + idx_hash_pair.first + "]" + width2range(data_width)
+           << " $end" << std::endl;
+  }
+  // let's go for the submodules
+  for (auto pos = scope->subscopes.begin() ; pos != scope->subscopes.end() ; ++ pos) {
+    fout << "$scope module " << pos->first << " $end" << std::endl;
+    dump_current_scope(fout, &(pos->second));
+    fout << "$upscope $end" << std::endl;
+  }
+} // end of dump_current_scope
+
+void VCDWitnessPrinter::DumpScopes(std::ostream & fout) const {
+  dump_current_scope(fout, &root_scope_);
+}
+
+void VCDWitnessPrinter::GenHeader(std::ostream & fout) const {
+  fout << "$date" << std::endl;
+  {
+    char buffer [100];
+    time_t rawtime;
+    struct tm * timeinfo;
+    time (&rawtime);
+    timeinfo = localtime (&rawtime);
+    if ( strftime(buffer, 100, date_time_format, timeinfo) == 0)
+      throw CosaException("Bug: time2string conversion failed.");
+    fout << buffer << std::endl;
+  }
+  fout << "$end" << std::endl;
+  fout << "$version CoSA2 $end" << std::endl;
+  fout << "$timescale 1 ns $end" << std::endl;
+  DumpScopes(fout);
+  fout << "$enddefinitions $end" << std::endl;
+} // end of GenHeader
+
+void VCDWitnessPrinter::dump_all(const smt::UnorderedTermMap & valmap,
+  std::unordered_map<std::string, std::string> & valbuf,
+  uint64_t t, std::ostream & fout) const {
+  for (auto && sig_bv_ptr : allsig_bv_) {
+    auto pos = valmap.find(sig_bv_ptr->ast);
+    if (pos == valmap.end()) {
+      logger.log(0, "missing value in provided trace @{}: {}" ,
```

Comment:
Sorry, one more thing: could we make this a level 1 log message please?
```suggestion
      logger.log(1, "missing value in provided trace @{}: {}" ,
```